File: agenkit-zig/src/middleware/rate_limiter.zig
/// Rate limiting middleware using token bucket algorithm
///
/// The token bucket algorithm allows for smooth rate limiting with burst capacity:
/// - Tokens are added to the bucket at a constant rate
/// - Each request consumes tokens from the bucket
/// - If insufficient tokens are available, the request waits or is rejected
/// - Burst capacity allows temporary spikes in traffic
///
/// This is useful for:
/// - Protecting downstream services from overload
/// - Complying with API rate limits (e.g., OpenAI: 3500 RPM)
/// - Fair resource allocation across tenants
/// - Cost control
///
/// Example:
/// ```zig
/// const config = RateLimiterConfig{
///     .rate = 100.0,              // 100 tokens per second
///     .capacity = 200,            // Allow bursts up to 200 requests
///     .tokens_per_request = 1,    // 1 token per request
/// };
///
/// var limiter = try RateLimiterDecorator.init(allocator, base_agent, config);
/// defer limiter.deinit();
///
/// const result = try limiter.agent().process(message);
/// const metrics = limiter.metrics();
/// ```
const std = @import("std");
const agksync = @import("../sync_compat.zig");
const agktime = @import("../time_compat.zig");
const json = std.json;
const Agent = @import("../agent.zig").Agent;
const AgentError = @import("../agent.zig").AgentError;
const StreamCallbacks = @import("../agent.zig").StreamCallbacks;
const Result = @import("../agent.zig").Result;
const Message = @import("../message.zig").Message;
const IntrospectionResult = @import("../introspection.zig").IntrospectionResult;
const createDefaultIntrospectionResult = @import("../introspection.zig").createDefaultIntrospectionResult;
const Allocator = std.mem.Allocator;

/// Rate limiter error
pub const RateLimitError = error{
    RateLimitExceeded,
};

/// Configuration for rate limiter behavior
pub const RateLimiterConfig = struct {
    /// Number of tokens added per second
    /// Default: 10.0
    rate: f64 = 10.0,

    /// Maximum burst capacity (maximum tokens in bucket)
    /// Default: 10
    capacity: u32 = 10,

    /// Number of tokens consumed per request
    /// Default: 1
    tokens_per_request: u32 = 1,

    /// Maximum time to wait for tokens in milliseconds (null = wait indefinitely)
    /// Prevents requests from waiting too long when rate limiter is heavily loaded
    /// Default: null (wait indefinitely)
    max_wait_timeout_ms: ?u64 = null,

    /// Validate configuration
    pub fn validate(self: RateLimiterConfig) !void {
        if (self.rate <= 0.0) {
            return error.InvalidConfig;
        }
        if (self.capacity < 1) {
            return error.InvalidConfig;
        }
        if (self.tokens_per_request < 1) {
            return error.InvalidConfig;
        }
        if (self.tokens_per_request > self.capacity) {
            return error.InvalidConfig;
        }
        if (self.max_wait_timeout_ms) |timeout| {
            if (timeout == 0) {
                return error.InvalidConfig;
            }
        }
    }
};

/// Metrics tracked by rate limiter middleware
pub const RateLimiterMetrics = struct {
    /// Total number of requests
    total_requests: u64 = 0,

    /// Number of allowed requests
    allowed_requests: u64 = 0,

    /// Number of rejected requests
    rejected_requests: u64 = 0,

    /// Total time spent waiting for tokens (milliseconds)
    total_wait_time_ms: u64 = 0,

    /// Current number of tokens in bucket
    current_tokens: f64 = 0.0,

    /// Average wait time per request (milliseconds)
    pub fn avgWaitTime(self: *const RateLimiterMetrics) ?f64 {
        if (self.total_requests == 0) return null;
        return @as(f64, @floatFromInt(self.total_wait_time_ms)) / @as(f64, @floatFromInt(self.total_requests));
    }

    /// Rejection rate (percentage)
    pub fn rejectionRate(self: *const RateLimiterMetrics) ?f64 {
        if (self.total_requests == 0) return null;
        return (@as(f64, @floatFromInt(self.rejected_requests)) / @as(f64, @floatFromInt(self.total_requests))) * 100.0;
    }

    /// Create a snapshot of metrics
    pub fn snapshot(self: *const RateLimiterMetrics) RateLimiterMetrics {
        return RateLimiterMetrics{
            .total_requests = self.total_requests,
            .allowed_requests = self.allowed_requests,
            .rejected_requests = self.rejected_requests,
            .total_wait_time_ms = self.total_wait_time_ms,
            .current_tokens = self.current_tokens,
        };
    }
};

/// Rate limiter decorator - wraps an agent with rate limiting protection
pub const RateLimiterDecorator = struct {
    allocator: Allocator,
    inner_agent: Agent,
    config: RateLimiterConfig,
    metrics_data: RateLimiterMetrics,
    tokens: f64,
    last_update_ms: i64,
    mutex: agksync.Mutex,

    pub fn init(allocator: Allocator, inner_agent: Agent, config: RateLimiterConfig) !*RateLimiterDecorator {
        // Validate configuration
        try config.validate();

        const now_ms = agktime.milliTimestamp();
        const initial_tokens = @as(f64, @floatFromInt(config.capacity));

        const self = try allocator.create(RateLimiterDecorator);
        self.* = RateLimiterDecorator{
            .allocator = allocator,
            .inner_agent = inner_agent,
            .config = config,
            .metrics_data = RateLimiterMetrics{ .current_tokens = initial_tokens },
            .tokens = initial_tokens,
            .last_update_ms = now_ms,
            .mutex = agksync.Mutex{},
        };
        return self;
    }

    pub fn agent(self: *RateLimiterDecorator) Agent {
        return Agent{
            .ptr = self,
            .vtable = &.{
                .name = nameImpl,
                .capabilities = capabilitiesImpl,
                .process = processImpl,
                .process_stream = processStreamImpl,
                .introspect = introspectImpl,
                .deinit = deinitImpl,
            },
        };
    }

    /// Get current metrics (thread-safe)
    pub fn metrics(self: *RateLimiterDecorator) RateLimiterMetrics {
        self.mutex.lock();
        defer self.mutex.unlock();
        return self.metrics_data.snapshot();
    }

    /// Refill tokens based on elapsed time
    fn refillTokens(self: *RateLimiterDecorator) void {
        const now_ms = agktime.milliTimestamp();
        const elapsed_ms = now_ms - self.last_update_ms;
        const elapsed_sec = @as(f64, @floatFromInt(elapsed_ms)) / 1000.0;

        // Add tokens based on elapsed time
        const tokens_to_add = elapsed_sec * self.config.rate;
        const capacity_f64 = @as(f64, @floatFromInt(self.config.capacity));
        self.tokens = @min(self.tokens + tokens_to_add, capacity_f64);
        self.last_update_ms = now_ms;

        // Update metrics
        self.metrics_data.current_tokens = self.tokens;
    }

    /// Acquire tokens from the bucket.
    ///
    /// Wait-and-retry loop. A single wait is not enough, for two independent
    /// reasons, and the old code returned an error from a branch it labelled
    /// "should not happen" when either bit:
    ///
    ///   1. refillTokens credits elapsed*rate from the wall clock, and
    ///      wait_time_ms truncates toward zero, so the sleep can return having
    ///      credited marginally less than tokens_needed.
    ///   2. The mutex is released across the sleep, so another thread can take
    ///      it first and drain the tokens this caller was waiting for -- a lost
    ///      wakeup. The equivalent Python path raised on every run with
    ///      capacity=1 and 8 concurrent callers (#750), despite a wait budget.
    ///
    /// This previously papered over (1) with an epsilon on the comparison, which
    /// narrowed the jitter window, did nothing for (2), and let a caller take
    /// tokens it had not earned. max_wait_timeout_ms still bounds the total,
    /// measured against cumulative wait, so this cannot spin forever when the
    /// configured rate genuinely cannot satisfy the request.
    fn acquireTokens(self: *RateLimiterDecorator, tokens_needed: u32, wait: bool) !void {
        const tokens_needed_f64 = @as(f64, @floatFromInt(tokens_needed));
        var total_waited_ms: u64 = 0;

        while (true) {
            self.mutex.lock();
            self.refillTokens();

            if (self.tokens >= tokens_needed_f64) {
                // Sufficient tokens available
                self.tokens -= tokens_needed_f64;
                self.metrics_data.current_tokens = self.tokens;
                if (total_waited_ms > 0) {
                    self.metrics_data.total_wait_time_ms += total_waited_ms;
                }
                self.mutex.unlock();
                return;
            }

            if (!wait) {
                // Insufficient tokens and not waiting
                self.mutex.unlock();
                return RateLimitError.RateLimitExceeded;
            }

            // Calculate wait time for the outstanding deficit. Floored at 1ms:
            // a sub-millisecond residual would otherwise truncate to a
            // zero-length sleep and spin hot until the wall clock caught up.
            const tokens_deficit = tokens_needed_f64 - self.tokens;
            const wait_time_sec = tokens_deficit / self.config.rate;
            var wait_time_ms = @as(u64, @intFromFloat(wait_time_sec * 1000.0));
            if (wait_time_ms < 1) {
                wait_time_ms = 1;
            }

            // Check the cumulative wait against max_wait_timeout so repeated
            // short retries cannot outlast the budget.
            if (self.config.max_wait_timeout_ms) |max_timeout| {
                if (total_waited_ms + wait_time_ms > max_timeout) {
                    self.mutex.unlock();
                    return RateLimitError.RateLimitExceeded;
                }
            }

            self.mutex.unlock();

            // Wait outside the lock to allow other operations
            agktime.sleep(wait_time_ms * std.time.ns_per_ms);
            total_waited_ms += wait_time_ms;
        }
    }

    fn nameImpl(ptr: *anyopaque) []const u8 {
        const self: *RateLimiterDecorator = @ptrCast(@alignCast(ptr));
        return self.inner_agent.name();
    }

    fn capabilitiesImpl(ptr: *anyopaque, allocator: Allocator) Allocator.Error![]const []const u8 {
        const self: *RateLimiterDecorator = @ptrCast(@alignCast(ptr));
        return self.inner_agent.capabilities(allocator);
    }

    fn processImpl(ptr: *anyopaque, message: Message) AgentError!Result {
        const self: *RateLimiterDecorator = @ptrCast(@alignCast(ptr));

        // Track request
        self.mutex.lock();
        self.metrics_data.total_requests += 1;
        self.mutex.unlock();

        // Acquire tokens (wait=true)
        self.acquireTokens(self.config.tokens_per_request, true) catch {
            self.mutex.lock();
            self.metrics_data.rejected_requests += 1;
            self.mutex.unlock();
            return AgentError.Cancelled; // Map RateLimitExceeded to Cancelled
        };

        // Token acquired successfully
        self.mutex.lock();
        self.metrics_data.allowed_requests += 1;
        self.mutex.unlock();

        // Process request
        return self.inner_agent.process(message);
    }

    fn introspectImpl(ptr: *anyopaque, allocator: Allocator) Allocator.Error!IntrospectionResult {
        const self: *RateLimiterDecorator = @ptrCast(@alignCast(ptr));

        // Get introspection from inner agent
        const inner_result = try self.inner_agent.introspect(allocator);

        // Add rate limiter metrics to metadata
        const metrics_snapshot = self.metrics();

        var metadata_map = json.ObjectMap.empty;
        errdefer metadata_map.deinit(allocator);

        // Add metrics as metadata
        try metadata_map.put(allocator, "total_requests", json.Value{ .integer = @intCast(metrics_snapshot.total_requests) });
        try metadata_map.put(allocator, "allowed_requests", json.Value{ .integer = @intCast(metrics_snapshot.allowed_requests) });
        try metadata_map.put(allocator, "rejected_requests", json.Value{ .integer = @intCast(metrics_snapshot.rejected_requests) });
        try metadata_map.put(allocator, "total_wait_time_ms", json.Value{ .integer = @intCast(metrics_snapshot.total_wait_time_ms) });
        try metadata_map.put(allocator, "current_tokens", json.Value{ .float = metrics_snapshot.current_tokens });

        if (metrics_snapshot.avgWaitTime()) |avg| {
            try metadata_map.put(allocator, "avg_wait_time_ms", json.Value{ .float = avg });
        }
        if (metrics_snapshot.rejectionRate()) |rate| {
            try metadata_map.put(allocator, "rejection_rate", json.Value{ .float = rate });
        }

        // Add configuration
        try metadata_map.put(allocator, "rate", json.Value{ .float = self.config.rate });
        try metadata_map.put(allocator, "capacity", json.Value{ .integer = @intCast(self.config.capacity) });
        try metadata_map.put(allocator, "tokens_per_request", json.Value{ .integer = @intCast(self.config.tokens_per_request) });

        // Merge with inner metadata (if it's an object)
        if (inner_result.metadata == .object) {
            var inner_iter = inner_result.metadata.object.iterator();
            while (inner_iter.next()) |entry| {
                const key = entry.key_ptr.*;
                if (!std.mem.startsWith(u8, key, "rate_")) {
                    try metadata_map.put(allocator, key, entry.value_ptr.*);
                }
            }
        }

        return IntrospectionResult{
            .allocator = allocator,
            .timestamp = agktime.timestamp(),
            .agent_name = inner_result.agent_name,
            .capabilities = inner_result.capabilities,
            .memory_state = inner_result.memory_state,
            .internal_state = inner_result.internal_state,
            .metadata = json.Value{ .object = metadata_map },
        };
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *RateLimiterDecorator = @ptrCast(@alignCast(ptr));
        // Don't deinit inner agent - caller manages it
        self.allocator.destroy(self);
    }
};

// Tests
const testing = std.testing;

fn processStreamImpl(ptr: *anyopaque, message: Message, callbacks: StreamCallbacks) AgentError!void {
    _ = ptr;
    _ = message;
    callbacks.onError(AgentError.NotImplemented);
}

test "RateLimiterConfig validation" {
    var config = RateLimiterConfig{};
    try config.validate();

    // Invalid: rate <= 0
    config.rate = 0.0;
    try testing.expectError(error.InvalidConfig, config.validate());
    config.rate = -1.0;
    try testing.expectError(error.InvalidConfig, config.validate());
    config.rate = 10.0;

    // Invalid: capacity < 1
    config.capacity = 0;
    try testing.expectError(error.InvalidConfig, config.validate());
    config.capacity = 10;

    // Invalid: tokens_per_request < 1
    config.tokens_per_request = 0;
    try testing.expectError(error.InvalidConfig, config.validate());
    config.tokens_per_request = 1;

    // Invalid: tokens_per_request > capacity
    config.tokens_per_request = 20;
    config.capacity = 10;
    try testing.expectError(error.InvalidConfig, config.validate());
}

test "RateLimiterMetrics calculations" {
    var metrics = RateLimiterMetrics{
        .total_requests = 10,
        .allowed_requests = 8,
        .rejected_requests = 2,
        .total_wait_time_ms = 5000,
    };

    // Average wait time
    const avg = metrics.avgWaitTime().?;
    try testing.expectApproxEqRel(@as(f64, 500.0), avg, 0.01);

    // Rejection rate
    const rejection_rate = metrics.rejectionRate().?;
    try testing.expectApproxEqRel(@as(f64, 20.0), rejection_rate, 0.01);
}

test "RateLimiterMetrics snapshot" {
    var metrics = RateLimiterMetrics{
        .total_requests = 5,
        .allowed_requests = 4,
        .rejected_requests = 1,
        .total_wait_time_ms = 1000,
        .current_tokens = 5.5,
    };

    const snapshot = metrics.snapshot();
    try testing.expectEqual(@as(u64, 5), snapshot.total_requests);
    try testing.expectEqual(@as(u64, 4), snapshot.allowed_requests);
    try testing.expectEqual(@as(u64, 1), snapshot.rejected_requests);
    try testing.expectEqual(@as(u64, 1000), snapshot.total_wait_time_ms);
    try testing.expectApproxEqRel(@as(f64, 5.5), snapshot.current_tokens, 0.01);
}

File: agenkit-rust/src/middleware/rate_limiter.rs
//! Rate limiter middleware using token bucket algorithm.
//!
//! Controls the rate of requests to prevent overwhelming downstream services,
//! respecting API rate limits, and ensuring fair resource allocation.
//!
//! # Algorithm: Token Bucket
//!
//! The token bucket algorithm maintains a bucket of tokens that refill at a constant rate:
//! - Each request consumes one token
//! - Tokens refill at `tokens_per_second` rate
//! - Maximum bucket capacity is `capacity`
//! - Requests wait if no tokens available (up to max_wait_time)
//!
//! **Example**: 10 tokens/second, capacity=20
//! - Allows bursts up to 20 requests instantly
//! - Sustained rate limited to 10 requests/second
//! - Smooths out traffic spikes
//!
//! # When to Use
//!
//! - **API rate limits**: Stay within provider limits (OpenAI, Anthropic, etc.)
//! - **Resource protection**: Prevent overwhelming downstream services
//! - **Cost control**: Limit spend on pay-per-request APIs
//! - **Fair sharing**: Ensure fair resource allocation across clients
//!
//! # When NOT to Use
//!
//! - **No rate limits**: When downstream has no rate constraints
//! - **Single request**: For one-off operations
//! - **Already limited**: When using provider's native rate limiting
//!
//! # Example
//!
//! ```rust
//! use agenkit::middleware::{RateLimiterMiddleware, RateLimiterConfig};
//! use agenkit::core::{Agent, Message};
//! use std::time::Duration;
//!
//! # async fn example() {
//! # struct MyAgent;
//! # #[async_trait::async_trait]
//! # impl Agent for MyAgent {
//! #     fn name(&self) -> &str { "test" }
//! #     async fn process(&self, msg: Message) -> Result<Message, agenkit::core::AgentError> {
//! #         Ok(Message::with_text("assistant", "ok"))
//! #     }
//! # }
//! let agent = MyAgent;
//!
//! let config = RateLimiterConfig::builder()
//!     .tokens_per_second(10.0)  // 10 requests/second
//!     .capacity(20.0)            // Allow bursts up to 20
//!     .max_wait_time(Duration::from_secs(5))
//!     .build();
//!
//! let rl_agent = RateLimiterMiddleware::new(agent, config);
//!
//! // Requests are rate-limited to 10/second
//! let msg = Message::with_text("user", "Hello");
//! let response = rl_agent.process(msg).await;
//! # }
//! ```

use crate::core::{Agent, AgentError, IntrospectionResult, Message};
use async_trait::async_trait;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::Mutex;

/// Configuration for rate limiter middleware.
#[derive(Debug, Clone)]
pub struct RateLimiterConfig {
    /// Tokens refilled per second (rate limit).
    /// Default: 10.0 tokens/second
    pub tokens_per_second: f64,

    /// Maximum bucket capacity (allows bursts).
    /// Default: Same as tokens_per_second
    pub capacity: f64,

    /// Maximum time to wait for a token.
    /// Default: 30 seconds
    pub max_wait_time: Duration,
}

impl Default for RateLimiterConfig {
    fn default() -> Self {
        Self {
            tokens_per_second: 10.0,
            capacity: 10.0,
            max_wait_time: Duration::from_secs(30),
        }
    }
}

impl RateLimiterConfig {
    /// Create a new builder for RateLimiterConfig.
    pub fn builder() -> RateLimiterConfigBuilder {
        RateLimiterConfigBuilder::default()
    }
}

/// Builder for RateLimiterConfig.
#[derive(Debug, Default)]
pub struct RateLimiterConfigBuilder {
    tokens_per_second: Option<f64>,
    capacity: Option<f64>,
    max_wait_time: Option<Duration>,
}

impl RateLimiterConfigBuilder {
    /// Set tokens per second (rate limit).
    pub fn tokens_per_second(mut self, rate: f64) -> Self {
        self.tokens_per_second = Some(rate);
        self
    }

    /// Set bucket capacity (burst size).
    pub fn capacity(mut self, capacity: f64) -> Self {
        self.capacity = Some(capacity);
        self
    }

    /// Set maximum wait time.
    pub fn max_wait_time(mut self, time: Duration) -> Self {
        self.max_wait_time = Some(time);
        self
    }

    /// Build the RateLimiterConfig.
    pub fn build(self) -> RateLimiterConfig {
        let default = RateLimiterConfig::default();
        let tokens_per_second = self.tokens_per_second.unwrap_or(default.tokens_per_second);
        RateLimiterConfig {
            tokens_per_second,
            capacity: self.capacity.unwrap_or(tokens_per_second), // Default capacity = rate
            max_wait_time: self.max_wait_time.unwrap_or(default.max_wait_time),
        }
    }
}

/// Metrics for rate limiter middleware.
#[derive(Debug, Clone, Default)]
pub struct RateLimiterMetrics {
    /// Total number of requests processed.
    pub total_requests: u64,

    /// Number of requests that were allowed immediately.
    pub allowed_requests: u64,

    /// Number of requests that were rejected (exceeded max wait time).
    pub rejected_requests: u64,

    /// Number of requests that had to wait for tokens.
    pub waited_requests: u64,

    /// Total time spent waiting for tokens (in milliseconds).
    pub total_wait_time_ms: u64,

    /// Current number of tokens in the bucket.
    pub current_tokens: f64,
}

impl RateLimiterMetrics {
    /// Calculate average wait time per request (in milliseconds).
    pub fn avg_wait_time_ms(&self) -> f64 {
        if self.total_requests == 0 {
            0.0
        } else {
            self.total_wait_time_ms as f64 / self.total_requests as f64
        }
    }

    /// Calculate wait rate (percentage of requests that waited).
    pub fn wait_rate(&self) -> f64 {
        if self.total_requests == 0 {
            0.0
        } else {
            (self.waited_requests as f64 / self.total_requests as f64) * 100.0
        }
    }

    /// Calculate rejection rate (percentage of requests rejected).
    pub fn rejection_rate(&self) -> f64 {
        if self.total_requests == 0 {
            0.0
        } else {
            (self.rejected_requests as f64 / self.total_requests as f64) * 100.0
        }
    }
}

/// Internal state for rate limiter.
#[derive(Debug)]
struct RateLimiterState {
    tokens: f64,
    last_refill: Instant,
    metrics: RateLimiterMetrics,
}

impl RateLimiterState {
    fn new(initial_tokens: f64) -> Self {
        Self {
            tokens: initial_tokens,
            last_refill: Instant::now(),
            metrics: RateLimiterMetrics {
                current_tokens: initial_tokens,
                ..Default::default()
            },
        }
    }

    /// Refill tokens based on elapsed time.
    fn refill(&mut self, tokens_per_second: f64, capacity: f64) {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last_refill).as_secs_f64();
        let new_tokens = elapsed * tokens_per_second;

        self.tokens = (self.tokens + new_tokens).min(capacity);
        self.last_refill = now;
        self.metrics.current_tokens = self.tokens;
    }

    /// Try to consume a token.
    fn try_consume(&mut self) -> bool {
        if self.tokens >= 1.0 {
            self.tokens -= 1.0;
            true
        } else {
            false
        }
    }

    /// Calculate wait time until next token is available.
    fn time_until_token(&self, tokens_per_second: f64) -> Duration {
        if self.tokens >= 1.0 {
            Duration::ZERO
        } else {
            let tokens_needed = 1.0 - self.tokens;
            let seconds = tokens_needed / tokens_per_second;
            Duration::from_secs_f64(seconds)
        }
    }
}

/// Rate limiter middleware using token bucket algorithm.
///
/// Controls request rate to prevent overwhelming downstream services,
/// respecting API rate limits, and ensuring predictable resource usage.
pub struct RateLimiterMiddleware<A: Agent> {
    inner: A,
    config: RateLimiterConfig,
    state: Arc<Mutex<RateLimiterState>>,
}

impl<A: Agent> RateLimiterMiddleware<A> {
    /// Create a new rate limiter middleware with the given agent and configuration.
    pub fn new(agent: A, config: RateLimiterConfig) -> Self {
        Self {
            inner: agent,
            state: Arc::new(Mutex::new(RateLimiterState::new(config.capacity))),
            config,
        }
    }

    /// Create a new rate limiter middleware with default configuration.
    pub fn with_defaults(agent: A) -> Self {
        Self::new(agent, RateLimiterConfig::default())
    }

    /// Get current rate limiter metrics.
    pub async fn get_metrics(&self) -> RateLimiterMetrics {
        let state = self.state.lock().await;
        state.metrics.clone()
    }
}

#[async_trait]
impl<A: Agent> Agent for RateLimiterMiddleware<A> {
    fn name(&self) -> &str {
        self.inner.name()
    }

    fn capabilities(&self) -> Vec<String> {
        self.inner.capabilities()
    }

    fn introspect(&self) -> IntrospectionResult {
        let mut result = self.inner.introspect();
        result.metadata.insert(
            "middleware".to_string(),
            serde_json::json!("rate_limiter"),
        );
        result.metadata.insert(
            "rate_limiter_config".to_string(),
            serde_json::json!({
                "tokens_per_second": self.config.tokens_per_second,
                "capacity": self.config.capacity,
                "max_wait_time_ms": self.config.max_wait_time.as_millis(),
            }),
        );
        result
    }

    async fn process(&self, message: Message) -> Result<Message, AgentError> {
        #[cfg(feature = "native")]
        {
            let start_wait = Instant::now();
            let mut waited = false;

            // Track total requests
            {
                let mut state = self.state.lock().await;
                state.metrics.total_requests += 1;
            }

            loop {
                let mut state = self.state.lock().await;

                // Refill tokens based on elapsed time
                state.refill(self.config.tokens_per_second, self.config.capacity);

                // Try to consume a token
                if state.try_consume() {
                    // Track metrics for successful acquisition
                    if waited {
                        state.metrics.waited_requests += 1;
                        let wait_time_ms = start_wait.elapsed().as_millis() as u64;
                        state.metrics.total_wait_time_ms += wait_time_ms;
                    } else {
                        state.metrics.allowed_requests += 1;
                    }
                    drop(state); // Release lock before calling inner agent
                    return self.inner.process(message).await;
                }

                // Check if we've exceeded max wait time
                if start_wait.elapsed() >= self.config.max_wait_time {
                    state.metrics.rejected_requests += 1;
                    return Err(AgentError::ProcessingError(
                        "rate limit exceeded: max wait time reached".to_string(),
                    ));
                }

                // Calculate wait time
                let wait_time = state.time_until_token(self.config.tokens_per_second);
                let remaining_time = self
                    .config
                    .max_wait_time
                    .saturating_sub(start_wait.elapsed());
                let actual_wait = wait_time.min(remaining_time);

                drop(state); // Release lock while waiting

                if actual_wait > Duration::ZERO {
                    waited = true;
                    tokio::time::sleep(actual_wait).await;
                }
            }
        }

        #[cfg(feature = "wasm")]
        {
            // WASM doesn't support tokio::time::sleep directly
            // For now, just call the inner agent without rate limiting
            // TODO: Implement rate limiting for WASM
            self.inner.process(message).await
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicU32, Ordering};

    struct CountingAgent {
        attempts: Arc<AtomicU32>,
    }

    impl CountingAgent {
        fn new() -> Self {
            Self {
                attempts: Arc::new(AtomicU32::new(0)),
            }
        }

        fn attempt_count(&self) -> u32 {
            self.attempts.load(Ordering::SeqCst)
        }
    }

    #[async_trait]
    impl Agent for CountingAgent {
        fn name(&self) -> &str {
            "counting"
        }

        async fn process(&self, _message: Message) -> Result<Message, AgentError> {
            self.attempts.fetch_add(1, Ordering::SeqCst);
            Ok(Message::with_text("assistant", "success"))
        }
    }

    #[tokio::test]
    async fn test_rate_limiter_allows_within_limit() {
        let agent = CountingAgent::new();
        let config = RateLimiterConfig::builder()
            .tokens_per_second(10.0)
            .capacity(5.0)
            .build();

        let rl_agent = RateLimiterMiddleware::new(agent, config);

        let msg = Message::with_text("user", "test");

        // Should allow up to capacity immediately
        for _ in 0..5 {
            let result = rl_agent.process(msg.clone()).await;
            assert!(result.is_ok());
        }

        assert_eq!(rl_agent.inner.attempt_count(), 5);
    }

    #[tokio::test]
    async fn test_rate_limiter_waits_when_empty() {
        let agent = CountingAgent::new();
        let config = RateLimiterConfig::builder()
            .tokens_per_second(10.0) // 10 tokens/sec = 100ms/token
            .capacity(1.0)
            .build();

        let rl_agent = RateLimiterMiddleware::new(agent, config);

        let msg = Message::with_text("user", "test");

        // First request should succeed immediately
        let start = Instant::now();
        let result = rl_agent.process(msg.clone()).await;
        assert!(result.is_ok());

        // Second request should wait ~100ms for token
        let result = rl_agent.process(msg.clone()).await;
        let elapsed = start.elapsed();

        assert!(result.is_ok());
        assert!(elapsed >= Duration::from_millis(90)); // Allow some timing variance
        assert_eq!(rl_agent.inner.attempt_count(), 2);
    }

    #[tokio::test]
    async fn test_rate_limiter_fails_after_max_wait() {
        let agent = CountingAgent::new();
        let config = RateLimiterConfig::builder()
            .tokens_per_second(1.0) // Very slow
            .capacity(0.0) // No initial tokens
            .max_wait_time(Duration::from_millis(100))
            .build();

        let rl_agent = RateLimiterMiddleware::new(agent, config);

        let msg = Message::with_text("user", "test");
        let result = rl_agent.process(msg).await;

        assert!(result.is_err());
        assert!(matches!(
            result.unwrap_err(),
            AgentError::ProcessingError(_)
        ));
    }

    #[tokio::test]
    async fn test_rate_limiter_refills_over_time() {
        let agent = CountingAgent::new();
        let config = RateLimiterConfig::builder()
            .tokens_per_second(10.0)
            .capacity(2.0)
            .build();

        let rl_agent = RateLimiterMiddleware::new(agent, config);

        let msg = Message::with_text("user", "test");

        // Consume initial tokens
        let _ = rl_agent.process(msg.clone()).await;
        let _ = rl_agent.process(msg.clone()).await;

        // Wait for refill (200ms = 2 tokens at 10/sec)
        tokio::time::sleep(Duration::from_millis(200)).await;

        // Should have refilled ~2 tokens
        let result = rl_agent.process(msg.clone()).await;
        assert!(result.is_ok());
    }

    #[tokio::test]
    async fn test_introspect_includes_rate_limiter_metadata() {
        let agent = CountingAgent::new();
        let config = RateLimiterConfig::builder()
            .tokens_per_second(5.0)
            .capacity(10.0)
            .build();

        let rl_agent = RateLimiterMiddleware::new(agent, config);
        let result = rl_agent.introspect();

        assert_eq!(
            result.metadata.get("middleware"),
            Some(&serde_json::json!("rate_limiter"))
        );
        assert!(result.metadata.contains_key("rate_limiter_config"));
    }
}
